bindings/python/src/tokenizer.rs

```diff
@@ -259,8 +259,24 @@ impl<'s> From<TextInputSequence<'s>> for tk::InputSequence<'s> {
 struct PyArrayUnicode(Vec<String>);
 impl FromPyObject<'_> for PyArrayUnicode {
     fn extract(ob: &PyAny) -> PyResult<Self> {
-        let array = ob.downcast::<PyArray1<u8>>()?;
-        let arr = array.as_array_ptr();
+        if unsafe { npyffi::PyArray_Check(ob.py(), ob.as_ptr()) } == 0 {
+            return Err(exceptions::PyTypeError::new_err("Expected an np.array"));
+        }
+        let arr = ob.as_ptr() as *mut npyffi::PyArrayObject;
+        if unsafe { (*arr).nd } != 1 {
+            return Err(exceptions::PyTypeError::new_err(
+                "Expected a 1 dimensional np.array",
+            ));
+        }
+        if unsafe { (*arr).flags }
+            & (npyffi::NPY_ARRAY_C_CONTIGUOUS | npyffi::NPY_ARRAY_F_CONTIGUOUS)
+            == 0
+        {
+            return Err(exceptions::PyTypeError::new_err(
+                "Expected a contiguous np.array",
+            ));
+        }
+        let n_elem = unsafe { *(*arr).dimensions } as usize;
```

Comment:
I feel quite uncomfortable with adding new `unsafe` code.

It was there previously just next line, but I would rather keep those unsafe limited, and ideally having a safety note associated with them.

I think this is used for sending data with `numpy` arrays which is ease of use not the main entry point.

I will investigate a bit more this part on my end to understand what's happening too.

do you know why the approach used in `PyArrayStr` does not work ?

Reply:
As @adamreichold said before, before 0.16 rust-numpy (incorrectly) did not check element type and dimension when downcasting to arrays, thus it was working by accident.

Reply:
> It was there previously just next line, but I would rather keep those unsafe limited, and ideally having a safety note associated with them.

Those unsafe blocks are required to check the preconditions of the existing code. `unsafe` is also non-local w.r.t function bodies, it can only be encapsulated at the module level. So there is little difference whether a function contains one or N unsafe blocks, it needs to be verified completely in both cases. (As the existing issue of assuming contiguous array without checking for that highlights.)

> do you know why the approach used in PyArrayStr does not work ?

This is just not implemented in `rust-numpy` yet, c.f. https://github.com/PyO3/rust-numpy/issues/141

Reply:
> non-local w.r.t  + here is little difference whether a function contains one or N unsafe blocks,

Yes unsafe is tricky and non local, but just because it's non local does not mean that adding new locations doesn't have implications. 

I didn't write the first implementation, so I do not necessarily understand the implications of those unsafe calls, I need to dive anyway to understand what was there previously. Crossing the ffi boundary is unsafe and definitely needed if `rust-numpy` does not implement some functions for us. I would just much prefer understanding a bit what's going on and adding SAFETY notes so we can remove those unsafe calls later when `rust-numpy` adds more support and takes on those SAFETY problems itself (if it does).

Reply:
I dived a bit into this.

Current understanding:
- TextInputSequence seems to work fine without including any unsafe, even on `numpy` array types, while working on real numpy arrays, not sure how that casting works.
- for testing we should be careful that `numpy.array` of strings can be either  `<UXX` when it uses pure bytes and aligns to the large string in the array I guess ?, and it uses `dtype('o')` when the sizes of lists cannot create a  tensor (which is non obvious when dtypes are not explicity mentionned in the numpy array creation), those cases are handled differently in rust.
- Current proposed unsafe does only ffi numpy boundary crossing (accessing the raw numpy array to recreate everything).
- Casting from `PyArray<_>` is not possible because the equivalent of `PyObject` (Which I guess is dtype('O'), right) does not exist yet for `<UXX` (And the unsafe code would basically be the `rust-numpy` implementation iiuc.)

I still need to investigate how `TextInputSequence` can bypass numpy casting altogether, and use purely `<Vec<&PyAny>>`. I tried doing the same for `PreTokenizedInputSequence` but somehow, I seemed to receive only the first part of the array, which is confusing (I must have done it wrong).

Does `TextInputSequence` create a whole lot of copies while succeeding to avoid casting to `PyArray` ?
Fixing/modifying this exceeds the scope of this PR, so let's not include this (unless we find a very simple solution) here.

Ideally, we could remove the `ffi` boundary, but so far it's OK.

1/ As for the current code, I see a lot of pointer dereferencing. Usually I think checking for the null pointer should be done ,I am guessing `PyArray_Check`,  we could probably add that for the first unsafe comment.

2/ Would it make sense to deference numpy array metadata only once, and then do all the error checking (Like we do for `let (type_num, elsize, alignment, data) = unsafe {` ) ? I think it makes more sense, but I am not sure what's considered better here.

3/ For the last part where we're actually recreating the strings, why do we need `PyUnicode_FromUnicode` (which is deprecated) and then casting it to `PyString` before converting it over to rust String. I think I saw a while back that `Python strings` !=  `Rust strings` but I cannot find it anymore, is that the case ? Otherwise since we have the byte array, can't we purely capture them as CString to String, or are they not necessarly `\0` terminated in the numpy byte format ?

Finally as a possible safety comments, I would see 3 sections.
```rust
/// SAFETY Making sure the pointer is a valid numpy array requires calling numpy C code

/// SAFETY Getting all the metadata about the numpy array to check its sanity

/// SAFETY Looking at the raw numpy data to create new owned Rust strings via copies (so it's safe afterwards).


Is my understanding correct ? Did I miss something ?

Reply:
> 3/ For the last part where we're actually recreating the strings, why do we need PyUnicode_FromUnicode (which is deprecated) and then casting it to PyString before converting it over to rust String. 

[numpy unicode](https://numpy.org/doc/stable/reference/c-api/dtype.html#c.NPY_UNICODE) is `UCS-4` while Rust string is UTF-8.

Reply:
Ok, diving was really important, as now much of your previous discussions make a lot more sense now !

Reply:
> Casting from PyArray<_> is not possible because the equivalent of PyObject (Which I guess is dtype('O'), right) does not exist yet for <UXX (And the unsafe code would basically be the rust-numpy implementation iiuc.)

Almost, the main problem is that we need to define which `T` in `PyArray<T>` safely maps to `<Un`. This might possibly be `[pyo3::ffi::Py_UCS4; n]`, but this might require const generics.